pkg/prove/querier.go

```diff
@@ -0,0 +1,56 @@
+package prove
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/celestiaorg/celestia-app/pkg/appconsts"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	"github.com/tendermint/tendermint/types"
+)
+
+const TxInclusionQueryPath = "txInclusionProof"
+
+// Querier defines the logic performed when the ABCI client using the Query
+// method with the custom prove.QueryPath. The index of the transaction being
+// proved must be appended to the path. The marshalled bytes of the transction
+// proof (tmproto.TxProof) are returned.
+//
+// example path for proving the third transaction in that block:
+// custom/txInclusionProof/3
+func QueryTxInclusionProof(_ sdk.Context, path []string, req abci.RequestQuery) ([]byte, error) {
+	// parse the index from the path
+	if len(path) != 1 {
+		return nil, fmt.Errorf("unexpected query path length %d", len(path))
+	}
+	index, err := strconv.ParseInt(path[0], 10, 64)
+	if err != nil {
+		return nil, err
+	}
+
+	// unmarshal the block data that is passed from the ABCI client
+	pbb := new(tmproto.Block)
+	err = pbb.Unmarshal(req.Data)
+	if err != nil {
+		return nil, fmt.Errorf("error reading block: %w", err)
+	}
+	data, err := types.DataFromProto(&pbb.Data)
+	if err != nil {
+		panic(fmt.Errorf("error from proto block: %w", err))
+	}
+
+	// create and marshal the tx inclusion proof, which we return in the form of []byte
+	txProof, err := TxInclusion(appconsts.DefaultCodec(), data, uint64(index))
```

Comment:
[Unrelated][Question]
In the docs of `TxInclusion(...)`:
```
// TxInclusion uses the provided block data to progressively generate rows
// of a data square, and then using those shares to creates nmt inclusion proofs
// It is possible that a transaction spans more than one row. In that case, we
// have to return two proofs.
func TxInclusion(codec rsmt2d.Codec, data types.Data, txIndex uint64) (types.TxProof, error) {
```
I checked the implementation, and if I understand right, it will generate a proof for each row a transaction spans on?
If so, why does it state in the docs:
```
In that case, we have to return two proofs.
```
and also:
```
var proofs []*tmproto.NMTProof  //nolint:prealloc // rarely will this contain more than a single proof
```

Reply:
They seem to match to me, but perhaps we can dig a bit deeper and clarify them. not sure I'm understanding the question properly, mind elaborating? which part doesn't match or make sense?

Reply:
The part I am not getting is the **two** proofs. Why not any number of proofs? Will the number of proofs be the same as the number of rows?

Reply:
true! yeah, we can change that, but all normal txs will only ever result in two proofs. You're correct tho that the implementation can handle more than two if that ever does happen (submitting a tx with a giiiiaaaannnnnntttttt memo maybe?)

Reply:
https://github.com/celestiaorg/celestia-app/pull/708/commits/26b8baa00ee372f609e8de7260b010f65c94b696

Reply:
Awesome,  thanks a lot